compiler/cli/src/org/jetbrains/kotlin/cli/jvm/plugins/DependencyShadingPluginClassLoader.kt

```diff
@@ -0,0 +1,68 @@
+
+import org.jetbrains.org.objectweb.asm.ClassReader
+import org.jetbrains.org.objectweb.asm.ClassWriter
+import org.jetbrains.org.objectweb.asm.commons.ClassRemapper
+import org.jetbrains.org.objectweb.asm.commons.Remapper
+import java.io.IOException
+import java.io.InputStream
+import java.net.URL
+import java.net.URLClassLoader
+import java.security.AccessControlContext
+import java.security.AccessController
+import java.security.PrivilegedActionException
+import java.security.PrivilegedExceptionAction
+
+/**
+ * DependencyShadingPluginClassLoader will load a plugin jar, applying the necessary shading as required
+ * by the variant of the Kotlin Compiler which is consuming the plugin.
+ */
+class DependencyShadingPluginClassLoader(urls: Array<URL>, parent: ClassLoader) : URLClassLoader(urls, parent) {
+
+    val accessControlContext: AccessControlContext
+    val bytecodeLoader = URLClassLoader(urls, null)
+
+    init {
+        accessControlContext = AccessController.getContext()
+    }
+
+    // True if the Kotlin Compiler loading this plugin is shaded, and thus the plugin jar should also be shaded
+    val shouldBeShaded = com.intellij.mock.MockProject::class.java.name.startsWith("org.jetbrains.kotlin.")
+
+    @Throws(ClassNotFoundException::class)
+    override fun findClass(name: String): Class<*>? {
+        return try {
+            AccessController.doPrivileged(
+                PrivilegedExceptionAction {
+                    val path = name.replace('.', '/') + ".class"
+                    val originalBytes = bytecodeLoader.getResourceAsStream(path)?.readBytes()
+
+                    if(originalBytes == null) throw ClassNotFoundException(name)
+
+                    try {
+                        val cw = ClassWriter(ClassWriter.COMPUTE_MAXS)
+
+                        val shader = ClassRemapper(cw, object : Remapper() {
+                            override fun map(internalName: String): String {
+                                if(shouldBeShaded && internalName.startsWith("com/intellij/"))
+                                    return "org/jetbrains/kotlin/"+internalName
+                                if(!shouldBeShaded && internalName.startsWith("org/jetbrains/kotlin/com/intellij/"))
+                                    return internalName.substringAfter("org/jetbrains/kotlin/")
+                                return internalName
+                            }
+                        })
+
+                        ClassReader(originalBytes).accept(shader, ClassReader.EXPAND_FRAMES)
+
+                        val shadedBytes = cw.toByteArray()
+                        defineClass(name, shadedBytes, 0, shadedBytes.size)
+                    } catch (e: IOException) {
+                        throw ClassNotFoundException(name, e)
+                    }
+                }, accessControlContext
+            )
+        } catch (e: PrivilegedActionException) {
+            throw ClassNotFoundException(name, e)
```

Comment:
It looks like it'd be better to `throw e.exception as ClassNotFoundException` since the contained exception is always a CNFE?

Reply:
Is `e.exception` guaranteed to be a `ClassNotFoundException`, now and in to the future?  For instance, if a `NullPointerException` was throwing, the cast would fail?  Certainly it isn't clear from the type system.  Also, rethrowing `e.exception` will cause the `PrivilegedActionException` to be eliminated from the trace, which can make the true control flow less discernible.  In general, I try to always wrap and never rethrow root causes.

My intuition is that throwing a new wrapper exception is more clear and provably correct.

But let me know if you'd prefer I cast e.exception instead, I'm happy to do whichever you prefer.

Reply:
I don't have a strong preference honestly. The javadoc of `PrivilegedActionException` and its usages hint to the fact that you're supposed to use its `exception` (=`cause`), and for example the implementation of `java.net.URLClassLoader` does this. But I guess it's fine either way, since the useful stack trace is not lost.

Reply:
Maybe something like this:
```
throw e.exception as? ClassNotFoundException ?: ClassNotFoundException(name, e)
```